### PR TITLE
DEV-18763: Add generic `ExtensionData`

### PR DESCRIPTION
**Pull request by JustAGabriel**
The `ExtensionData` mechanism was refactored to reduce code duplication and allow generic application.

### CODE REVIEW

**Comment by JustAGabriel** [comment]
Disclaimer: you actually have to thank @GodsBoss for making this work at all!

File: internal/unmappeddatamarshaller/unmappeddatamarshaller.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/hochfrequenz/go-bo4e/internal/jsonfieldnames"
+)
+
+type UnmappedData struct {
+	Data map[string]any
+}
```
Comment (hf-kklein) [inline internal/unmappeddatamarshaller/unmappeddatamarshaller.go:11]: could you add a little docstring on what's its purpose and so on?

**Review by hf-kklein — COMMENTED** [review]
I'll let @GodsBoss do the review :) Definitly better qualified

File: internal/unmappeddatamarshaller/unmappeddatamarshaller.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+func (umd *UnmappedData) storeUnmappedData(key string, value any) {
+	umd.Data[key] = value
+}
+
+func MarshalWithUnmappedData[T any](s T) (bytes []byte, err error) {
+	byteArr, err := json.Marshal(s)
```
[the remark this answers is not in the document]
Comment (GodsBoss) [reply internal/unmappeddatamarshaller/unmappeddatamarshaller.go:18]: Okay, that one was a bit tricky. You cannot directly shadow a generic type like this:
```
func Foo[T any](t T) {
    type S T
}
```
This only results in a compilation error `cannot use a type parameter as rhs in type declaration`.

But with a trick similar to the one used in `market_communication.GetAll()` it is possible to create a type derived from the generic type. Take a look at [this example](https://go.dev/play/p/4McaJ2l6EMQ) - The `UnmarshalJSON` method calls the `Unmarshal` function which in turn calls `json.Unmarshal` on the value. The derived type has no `UnmarshalJSON` method, so no infinite loop.

File: bo/businessobject.go [hunk 3]
```diff
@@ -17,27 +18,10 @@ type BusinessObject interface {
...
-		gob.DeleteExtensionData(key)
-	}
-}
-
-// DeleteExtensionData removes the given jsonKey from the Geschaeftsobjekt.ExtensionData
-func (gob *Geschaeftsobjekt) DeleteExtensionData(jsonKey string) {
-	if gob.ExtensionData != nil {
-		delete(gob.ExtensionData, jsonKey)
-	}
+	unmappeddatamarshaller.UnmappedData
```
Comment (GodsBoss) [inline bo/businessobject.go:21]: @JustAGabriel @hf-kklein I think we may get away with leaving `ExtensionData`. `unmappeddatamarshaller.UnmappedData` has only one method, `storeUnmappedData`, but that only assigns a value in a map and is used once, so we could inline that logic and remove the method. This leaves us with a struct wrapper around a map and we can pass the map directly to `unmappeddatamarshaller.UnmarshallWithUnmappedData`. Of course the logic in `unmappeddatamarshaller.MarshalWithUnmappedData` needs to take that change into account, too.

File: internal/unmappeddatamarshaller/unmappeddatamarshaller.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	if err != nil {
+		return
+	}
+
+	unmappedDataFieldNames, err := jsonfieldnames.Extract(UnmappedData{})
+	if err != nil {
+		return
+	} else if len(unmappedDataFieldNames) != 1 {
+		err = errors.New("expected exactly one field ('Data' in struct 'UnmappedData')")
```
Comment (GodsBoss) [inline internal/unmappeddatamarshaller/unmappeddatamarshaller.go:27]: Can this ever happen? It extracts the field names from `UnmappedData`, which is a concrete type. I proposed changes that would remove the type `UnmappedData` completely (see removal of `ExtensionData`), so this is probably not important.

**Review by GodsBoss — COMMENTED** [review]
Added a few comments.

File: internal/unmappeddatamarshaller/unmappeddatamarshaller.go (hunk 4 above)
Comment (JustAGabriel) [reply internal/unmappeddatamarshaller/unmappeddatamarshaller.go]: was removed

File: bo/bilanzierung_test.go [hunk 5]
```diff
@@ -83,8 +84,8 @@ func (s *Suite) Test_Bilanzierung_Deserializes_Unknown_Fields() {
 	var deserializedBilanzierung bo.Bilanzierung
 	err := json.Unmarshal([]byte(jsonString), &deserializedBilanzierung)
 	then.AssertThat(s.T(), err, is.Nil())
-	then.AssertThat(s.T(), deserializedBilanzierung.Bilanzkreis, is.EqualTo("THE0BFL002410004"))              // a "normal" property/field of Bilanzierung
-	then.AssertThat(s.T(), deserializedBilanzierung.ExtensionData["Lastprofil_Codeliste"], is.EqualTo("293")) // an extension data key
+	then.AssertThat(s.T(), deserializedBilanzierung.Bilanzkreis, is.EqualTo("THE0BFL002410004"))                  // a "normal" property/field of Bilanzierung
+	then.AssertThat(s.T(), deserializedBilanzierung.UnmappedData.Data["Lastprofil_Codeliste"], is.EqualTo("293")) // an extension data key
```
[the remark this answers is not in the document]
Comment (JustAGabriel) [reply bo/bilanzierung_test.go]: Was addressed. Please have another look.

File: internal/unmappeddatamarshaller/unmappeddatamarshaller.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+//			unmappedDataMap := value.(map[string]any)
+//			for k, v := range unmappedDataMap {
+//				structFields[k] = v
+//			}
+//			delete(structFields, fieldName)
+//		}
+//	}
+//
+//	return json.Marshal(structFields)
+//}
```
Comment (hf-kklein) [inline internal/unmappeddatamarshaller/unmappeddatamarshaller.go:57]: @JustAGabriel können wir das entfernen?

File: internal/unmappeddatamarshaller/unmappeddatamarshaller.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+				}
+				delete(structFields, fieldName)
+			}
+		}
+	}
+
+	return json.Marshal(structFields)
+}
+
+func UnmarshallWithUnmappedData[T any](targetStruct *T, unmappedDataInTargetStruct *ExtensionData, bytes []byte) (err error) {
```
Comment (hf-kklein) [inline internal/unmappeddatamarshaller/unmappeddatamarshaller.go:83]: magst du den beiden methoden noch einen docstring geben, der highlevel beschreibt, was sie tun?

**Review by hf-kklein — APPROVED** [review]

**Pull request merged**